### PR TITLE
feature: added ssl_certificiate_by_lua* directives to allow controlling downstream SSL handshakes with Lua

### PR DESCRIPTION
**Pull request by agentzh**
This pull request relies on the following nginx core patch:

https://github.com/openresty/ngx_openresty/blob/master/patches/nginx-1.9.7-ssl_cert_cb_yield.patch

(already proposed to the nginx-devel mailing list: http://mailman.nginx.org/pipermail/nginx-devel/2016-January/007748.html )

This pull request couples with the `ssl-cert-by-lua-2` branch of `lua-resty-core`. See the following pull request of `lua-resty-core` for the Lua API:

https://github.com/openresty/lua-resty-core/pull/12

Related new documentation:

https://github.com/openresty/lua-nginx-module/tree/ssl-cert-by-lua-2#ssl_certificate_by_lua_block

See #331 for more details on the requirements and development history.

### CODE REVIEW

**Comment by agentzh** [comment]
@lziest @lordnynex @bungle @doujiang24 @andrea-spoldi @moseleymark @marklkelly @yejingx and others (sorry if I didn't mention your names explicitly). Please review this PR for me :) Thanks!

File: patches/nginx-ssl-cert.patch [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ 
+     c->ssl->no_wait_shutdown = 1;
+diff --exclude '*~' '--exclude=*.swp' -upr a/src/event/ngx_event_openssl.h b/src/event/ngx_event_openssl.h
+--- a/src/event/ngx_event_openssl.h	2014-08-05 04:13:07.000000000 -0700
++++ b/src/event/ngx_event_openssl.h	2014-09-12 12:16:32.016208272 -0700
+@@ -56,6 +56,8 @@ typedef struct {
+     ngx_event_handler_pt        saved_read_handler;
+     ngx_event_handler_pt        saved_write_handler;
+ 
++    void                       *lua_ctx;  /* used by 3rd-party modules */
```
Comment (lziest) [inline patches/nginx-ssl-cert.patch]: so this can be removed now. right?

**Comment by lziest** [comment]
lgtm

**Comment by bungle** [comment]
Hi,

On ssl.t there is lines like this:

```
received: Server: nginx
```

Should it really be:

```
received: Server: openresty
```

**Comment by bungle** [comment]
If I change that received: Server thing, I still get a few errors:
https://gist.github.com/bungle/c969454d66bafbe9c046

Is it because of my configs or something?

**Comment by bungle** [comment]
I also did added some locals to ocsp.lua and ssl.lua here:
https://github.com/openresty/lua-resty-core/pull/13

File: patches/nginx-ssl-cert.patch (hunk 1 above)
Comment (agentzh) [inline patches/nginx-ssl-cert.patch]: @lziest Oh, this patch is already obsolete. I've just removed the whole patch file :) Thanks for the catch.

**Comment by agentzh** [comment]
@bungle The `Server` header thing is that I usually use NGINX instead of OpenResty to run those tests. In particular, I use the following script to build an NGINX to run these tests:

https://github.com/openresty/lua-nginx-module/blob/master/util/build2.sh

To run the tests on Mac OS X, we have to add 127.0.0.2 to the loopback device, via the following command:

```
sudo ifconfig lo0 alias 127.0.0.2 up
```

**Comment by agentzh** [comment]
@bungle The failed raw_server_addr tests on Mac OS X reveal a bug in my implementation. I'll commit a fix soon :) Thanks!

**Comment by agentzh** [comment]
Okay, this branch has just been merged into master. I'm closing this. Further issues should go to separate tickets :)

**Pull request closed without merging**